### PR TITLE
make timeout interfaces async

### PR DESCRIPTION
**Pull request by timbussmann**
This PR is part of the work to make the core async enabled. For more background read [Async/Await: It's time!](http://particular.net/blog/async-await-its-time) blogpost

* [x] IQueryTimeouts
* [x] PersistTimeouts

### CODE REVIEW

**Comment by timbussmann** [comment]
@danielmarbach could you review?

File: src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt [hunk 1]
```diff
@@ -2317,13 +2317,13 @@ namespace NServiceBus.Timeout.Core
 {
     public interface IPersistTimeouts
     {
-        void Add(NServiceBus.Timeout.Core.TimeoutData timeout, NServiceBus.Timeout.Core.TimeoutPersistenceOptions options);
-        void RemoveTimeoutBy(System.Guid sagaId, NServiceBus.Timeout.Core.TimeoutPersistenceOptions options);
-        bool TryRemove(string timeoutId, NServiceBus.Timeout.Core.TimeoutPersistenceOptions options, out NServiceBus.Timeout.Core.TimeoutData timeoutData);
+        System.Threading.Tasks.Task Add(NServiceBus.Timeout.Core.TimeoutData timeout, NServiceBus.Timeout.Core.TimeoutPersistenceOptions options);
+        System.Threading.Tasks.Task RemoveTimeoutBy(System.Guid sagaId, NServiceBus.Timeout.Core.TimeoutPersistenceOptions options);
+        System.Threading.Tasks.Task<NServiceBus.Timeout.Core.TimeoutData> TryRemove(string timeoutId, NServiceBus.Timeout.Core.TimeoutPersistenceOptions options);
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt]: With outbox I removed the `Try`. I think we should do the same here

File: src/NServiceBus.Core.Tests/Timeout/InMemoryTimeoutPersisterThreadTests.cs [hunk 2]
```diff
@@ -26,7 +26,7 @@ public void Run()
 
             for (var i = 0; i < 10; i++)
             {
-                var thread = new Thread(() => Runner(inMemoryTimeoutPersister, options));
+                var thread = new Thread(() => Runner(inMemoryTimeoutPersister, options).Wait());
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/Timeout/InMemoryTimeoutPersisterThreadTests.cs]: It is probably a none issue. But I don't like `Thread(() => Runner().Wait())`. We could just use `Task.Run`

File: src/NServiceBus.Core.Tests/Timeout/When_fetching_timeouts_from_storage.cs [hunk 3]
```diff
@@ -60,28 +62,26 @@ public void Should_set_the_next_run()
...
+            var nextChunk = await GetNextChunk();
 
-            var totalMilliseconds = (expected - nextTimeToRunQuery).Duration().TotalMilliseconds;
+            var totalMilliseconds = (expected - nextChunk.NextTimeToQuery).Duration().TotalMilliseconds;
 
             Assert.True(totalMilliseconds < 200);
         }
 
-        IEnumerable<Tuple<string, DateTime>> GetNextChunk()
+        Task<TimeoutsChunk> GetNextChunk()
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/Timeout/When_fetching_timeouts_from_storage.cs]: I'd prefer to inline this method

File: src/NServiceBus.Core.Tests/Timeout/When_removing_timeouts_from_the_storage.cs [hunk 4]
```diff
@@ -22,29 +22,28 @@ public void Setup()
...
             }
 
-            Assert.AreEqual(0, GetNextChunk().Count());
+            var secondChunk = await GetNextChunk();
+
+            Assert.AreEqual(0, secondChunk.DueTimeouts.Count());
         }
 
-        IEnumerable<Tuple<string, DateTime>> GetNextChunk()
+        Task<TimeoutsChunk> GetNextChunk()
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/Timeout/When_removing_timeouts_from_the_storage.cs]: Here as well

File: src/NServiceBus.Core.Tests/Timeout/InMemoryTimeoutPersisterThreadTests.cs (hunk 2 above)
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/Timeout/InMemoryTimeoutPersisterThreadTests.cs]: Yes we could, as I tried that. But the test is called ThreadTest so I'm not sure what the reason behind the test really is, so I left it as is.

File: src/NServiceBus.Core.Tests/Timeout/When_fetching_timeouts_from_storage.cs (hunk 3 above)
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/Timeout/When_fetching_timeouts_from_storage.cs]: fine with that :+1: . Overlooked these after removing the out parameter.

**Comment by danielmarbach** [comment]
@andreasohlund @timbussmann can you guys do a review?

@andreasohlund have a proper look at the pump changes

**Comment by danielmarbach** [comment]
Rebased. Green now

**Comment by danielmarbach** [comment]
@timbussmann and/or @andreasohlund can you check this and merge?

**Comment by danielmarbach** [comment]
I think we can make this

https://github.com/Particular/NServiceBus/blob/timeout-interfaces-async/src/NServiceBus.Core/Transports/Msmq/MessagePump.cs#L73

TaskScheduler.Default

thoughts?

**Comment by andreasohlund** [comment]
TaskScheduler.Default makes sense.

Would that include https://github.com/Particular/NServiceBus/blob/timeout-interfaces-async/src/NServiceBus.Core/Transports/Msmq/MessagePump.cs#L68 as well?

**Comment by timbussmann** [comment]
is there a chance that we "enqueue too many tasks" in case of a limited scheduler when having a lot of messages in the queue? Does the queue enumerator/peek only yield once per message?

**Comment by andreasohlund** [comment]
>is there a chance that we "enqueue too many tasks" in case of a limited scheduler when having a lot of messages in the queue (when using the default scheduler)? Does the queue enumerator/peek only yield once per message?

It needs to be double checked, but the enumerator will only enqueue one task per message, but I need to check what happens when it reaches the end of the queue

**Comment by danielmarbach** [comment]
@andreasohlund yes. I'll change

**Comment by danielmarbach** [comment]
@andreasohlund  ConcurrencyLimit added, Removed the cancellation of the pump task and make it not move into cancelled state. Then we can just await its completion

**Comment by danielmarbach** [comment]
@timbussmann please also review

**Pull request merged**